### PR TITLE
tolerate missing docker images

### PR DESCRIPTION
**Pull request by bparees**
fixes for https://bugzilla.redhat.com/show_bug.cgi?id=1231749

### CODE REVIEW

**Comment by bparees** [comment]
@csrwng ptal.

File: pkg/cmd/cli/cmd/newapp.go [hunk 1]
```diff
@@ -144,6 +144,7 @@ func NewCmdNewApplication(fullName string, f *clientcmd.Factory, out io.Writer)
 	cmd.Flags().BoolVar(&config.InsecureRegistry, "insecure-registry", false, "If true, indicates that the referenced Docker images are on insecure registries and should bypass certificate checking")
 	cmd.Flags().BoolVarP(&config.AsList, "list", "L", false, "List all local templates and image streams that can be used to create.")
 	cmd.Flags().BoolVarP(&config.AsSearch, "search", "S", false, "Search all templates, image streams, and Docker images that match the arguments provided.")
+	cmd.Flags().BoolVar(&config.AllowMissing, "allow-missing", false, "If true, indicates that referenced Docker images that cannot be found locally or in a registry should still be used.")
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newapp.go]: We likely want the same flag in cmd/newbuild.go

File: pkg/image/api/helper.go [hunk 2]
```diff
@@ -165,6 +165,21 @@ func (r DockerImageReference) String() string {
...
 }
 
+// StringNoTag converts a DockerImageReference to a Docker pull spec with no tag.
+func (r DockerImageReference) StringNoTag() string {
+	registry := r.Registry
+	if len(registry) > 0 {
+		registry += "/"
+	}
+
+	if len(r.Namespace) == 0 {
```
Comment (csrwng) [inline pkg/image/api/helper.go]: why not use a local var like you do for registry above? You're not modifying the ref because you're working on a copy of it, but if in the future we change the method receiver to a ptr, then you'd be inadvertently mutating the object by just getting its String

Comment (bparees) [inline pkg/image/api/helper.go]: this is copied directly from the existing String() method above...

Comment (csrwng) [inline pkg/image/api/helper.go]: I noticed :) We should change both. We should at least be internally consistent... if we use a local var for the registry then we should use one for the namespace. Or don't for both.

Comment (bparees) [inline pkg/image/api/helper.go]: updated.

**Comment by bparees** [comment]
[test]

File: pkg/cmd/cli/cmd/newapp.go (hunk 1 above)
Comment (bparees) [inline pkg/cmd/cli/cmd/newapp.go]: added.

File: pkg/generate/app/dockerimagelookup.go [hunk 3]
```diff
@@ -108,6 +112,13 @@ func (r DockerClientSearcher) Search(terms ...string) (ComponentMatches, error)
 		}
 
 		componentMatches = append(componentMatches, termMatches...)
+		if len(componentMatches) == 0 && r.AllowMissing {
```
Comment (csrwng) [inline pkg/generate/app/dockerimagelookup.go]: The searcher will also be used when you do a call with --search=true. I suppose in that case we wouldn't want the --allow-missing flag to be true. Maybe we should add validation to the flags in the command to make them mutually exclusive.

**Comment by csrwng** [comment]
@bparees just one more comment and then it lgtm

**Comment by csrwng** [comment]
LGTM

**Comment by bparees** [comment]
[merge]

**Comment by bparees** [comment]
missed bash completion updates.  [merge]

**Comment by bparees** [comment]
[merge]

**Comment by bparees** [comment]
[merge]

File: pkg/image/api/helper.go [hunk 4]
```diff
@@ -157,12 +157,28 @@ func (r DockerImageReference) String() string {
...
 
-	if len(r.Namespace) == 0 {
-		r.Namespace = DockerDefaultNamespace
+	namespace := r.Namespace
+	if len(namespace) == 0 {
+		namespace = DockerDefaultNamespace
+	}
+	namespace += "/"
+
+	return fmt.Sprintf("%s%s%s", registry, namespace, r.NameString())
```
Comment (mfojtik) [inline pkg/image/api/helper.go]: `return registry + namespace + r.NameString()` :-) but I'm too late

File: pkg/image/api/helper.go [hunk 5]
```diff
@@ -157,12 +157,28 @@ func (r DockerImageReference) String() string {
 		registry += "/"
 	}
 
-	if len(r.Namespace) == 0 {
-		r.Namespace = DockerDefaultNamespace
+	namespace := r.Namespace
```
Comment (smarterclayton) [inline pkg/image/api/helper.go]: Please undo this change - we explicitly do *not* manipulate these states during String, instead a caller should use `Minimal` or a new variant.

File: pkg/image/api/helper.go [hunk 6]
```diff
@@ -157,12 +157,28 @@ func (r DockerImageReference) String() string {
...
+	if len(namespace) == 0 {
+		namespace = DockerDefaultNamespace
+	}
+	namespace += "/"
+
+	return fmt.Sprintf("%s%s%s", registry, namespace, r.NameString())
+}
+
+// StringNoTag converts a DockerImageReference to a Docker pull spec with no tag.
+func (r DockerImageReference) StringNoTag() string {
```
Comment (smarterclayton) [inline pkg/image/api/helper.go]: This should be `NoTag` on the Reference, and then the user should print the tag.  We should never manipulate the reference during String generation.

Comment (smarterclayton) [inline pkg/image/api/helper.go]: This also needs test cases - please open a new PR that undoes these changes with the requested changes described above.

File: pkg/image/api/helper.go (hunk 5 above)
Comment (smarterclayton) [inline pkg/image/api/helper.go]: This method is fine, rather.  It's the subsequent method that needs to be reverted.

File: pkg/image/api/helper.go (hunk 6 above)
Comment (smarterclayton) [inline pkg/image/api/helper.go]: Nm, I'll just do it in my PR that overlaps this.

File: pkg/cmd/cli/cmd/newapp.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newapp.go]: `allow-missing` isn't a great name.  I'd prefer to be more descriptive here, especially since this flag isn't something a user is going to find automatically.  Probably should be `allow-missing-images` or `use-images-if-missing`

File: pkg/image/api/helper.go (hunk 6 above)
Comment (smarterclayton) [inline pkg/image/api/helper.go]: Changed in #4934

File: pkg/cmd/cli/cmd/newapp.go (hunk 1 above)
Comment (bparees) [inline pkg/cmd/cli/cmd/newapp.go]: the help explicilty tells the user to use this flag if they hit a situation where it would be relevant, so it's pretty discoverable imho.

Comment (bparees) [inline pkg/cmd/cli/cmd/newapp.go]: (i'm not against allow-missing-images, just felt really long)

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newapp.go]: Rule of thumb is going to be how often someone needs this, and the rarer it
is the longer the description is allowed to be.  I assume this is only in
very specialized cases, like I'm going to be pushing an image and I want to
create it later?  That seems in the 1-2% of the time range.

Comment (bparees) [inline pkg/cmd/cli/cmd/newapp.go]: Hopefully, yes, I'll do a pr to rename it today.

Ben Parees | OpenShift
